exonum/src/events/noise/wrapper.rs

```diff
@@ -126,35 +126,50 @@ impl NoiseWrapper {
     /// 4. Append all encrypted packets in corresponding order.
     /// 5. Write result message to `buf`
     pub fn encrypt_msg(&mut self, msg: &[u8], buf: &mut BytesMut) -> Result<Option<()>, io::Error> {
-        let mut len = 0_usize;
-        let mut encoded_message = vec![0_u8; 0];
+        let len = self.encrypted_msg_len(msg.len());
+        let mut encoded_message = Vec::with_capacity(len);
 
-        msg.chunks(NOISE_MAX_MESSAGE_LENGTH - TAG_LENGTH)
-            .for_each(|msg| {
-                let (written_bytes, written) = self.write(msg).unwrap();
-                encoded_message.extend_from_slice(&written);
-                len += written_bytes;
-            });
+        for msg in msg.chunks(MAX_MESSAGE_LENGTH - TAG_LENGTH) {
+            let written = self.write(msg)?;
+            encoded_message.extend_from_slice(&written);
+        }
 
         let mut msg_len_buf = vec![0_u8; NOISE_HEADER_LENGTH];
 
         LittleEndian::write_u32(&mut msg_len_buf, len as u32);
         let encoded_message = &encoded_message[0..len];
         msg_len_buf.extend_from_slice(encoded_message);
         buf.extend_from_slice(&msg_len_buf);
+
+        debug_assert_eq!(len, encoded_message.len());
         Ok(None)
     }
 
-    fn read(&mut self, input: &[u8], len: usize) -> Result<(usize, Vec<u8>), NoiseError> {
+    fn read(&mut self, input: &[u8], len: usize) -> Result<Vec<u8>, NoiseError> {
         let mut buf = vec![0_u8; len];
-        let len = self.session.read_message(input, &mut buf)?;
-        Ok((len, buf))
+        self.session.read_message(input, &mut buf)?;
+        Ok(buf)
     }
 
-    fn write(&mut self, msg: &[u8]) -> Result<(usize, Vec<u8>), NoiseError> {
-        let mut buf = vec![0_u8; NOISE_MAX_MESSAGE_LENGTH];
+    fn write(&mut self, msg: &[u8]) -> Result<Vec<u8>, NoiseError> {
+        let mut buf = vec![0_u8; MAX_MESSAGE_LENGTH];
         let len = self.session.write_message(msg, &mut buf)?;
-        Ok((len, buf))
+        buf.truncate(len);
+        Ok(buf)
+    }
+
+    // Each message consists of the payload and 16 bytes(`TAG_LENGTH`)
+    // of AEAD authentification data. Therefore to calculate an actual message
+    // length we need to substract `TAG_LENGTH` multiplied by messages count
+    // from `data.len()`.
+    fn decrypted_msg_len(&self, raw_message_len: usize) -> usize {
+        raw_message_len - TAG_LENGTH * (raw_message_len / MAX_MESSAGE_LENGTH)
```

Comment:
I'm not sure the math is correct here (is it covered by tests?). If each Noise message contains a tag, then the formula would be

```rust
len - TAG_LENGTH * ceil(len / MAX_MESSAGE_LENGTH)
```

(Is there `ceil` for integers in stdlib?)

Reply:
Your formula is right, 
```rust
raw_message_len - TAG_LENGTH * ((raw_message_len as f32 / MAX_MESSAGE_LENGTH as f32).ceil() as usize)
```
In the current implementation, we have an extra 16 bytes on the end of each message. But they are ignored because we specify length in a message header. I'll fix it in next PR. Tests also will be provided.